### PR TITLE
rustdoc: Put GenericArg::Type into a Box

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This is a small experiment to see how it'll impact performance (I think it'll increase memory usage a bit). I don't think it'll be worthwhile to merge it.

r? @camelid

### CODE REVIEW

**Comment by petrochenkov** [comment]
@GuillaumeGomez 
Meta: could you mark PR names with the code are that they touch, like "rustdoc: Put GenericArg::Type into a Box"?
Right now from the title it may look like the PR changes AST or HIR.

**Comment by GuillaumeGomez** [comment]
I'm so sorry I forgot again... >< I'll add it here but sorry for the useless noise...

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -1980,12 +1980,14 @@ impl Path {
...
     crate fn generics(&self) -> Option<Vec<&Type>> {
+        use std::ops::Deref;
+
         self.segments.last().and_then(|seg| {
             if let GenericArgs::AngleBracketed { ref args, .. } = seg.args {
                 Some(
                     args.iter()
                         .filter_map(|arg| match arg {
-                            GenericArg::Type(ty) => Some(ty),
+                            GenericArg::Type(ty) => Some(ty.deref()),
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1990]: I did write it `(**ty)` originally but it was still unhappy for some reason...

**Comment by camelid** [comment]
Hmm, the instruction count regressions are higher than I'd like ... but the memory improvement does seem nice. Ideally, we would try to shrink `Type` instead of just boxing it—though that's quite difficult. I'm ambivalent about this change; I'm not sure whether to merge it or close it.

File: src/librustdoc/clean/types.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/types.rs:1990]: Can you post the error then?

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1990]: Sure, will do that tomorrow.

**Comment by camelid** [comment]
Yeah, I've been meaning to get back to trying to shrink Type for a while now. If you want to look into it, can you try to get rid of https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/clean/types/enum.Type.html#variant.QPath.field.self_def_id? That's a hack that should be removed and takes up decent space.

**Comment by camelid** [comment]
My current thinking is the perf improvements aren't worth the regressions and increased code complexity (having to box and unbox). So I propose closing this PR.

**Comment by GuillaumeGomez** [comment]
Agreed!

**Pull request closed without merging**